runtime/src/main/java/org/corfudb/recovery/FastSmrMapsLoader.java

```diff
@@ -0,0 +1,328 @@
+
+    /** Extract log entries from logData and update the Corfu Objects
+     *
+     * @param logData LogData received from Corfu server.
+     */
+    private void processLogData(ILogData logData) {
+
+        LogEntry logEntry;
+        try {
+            logEntry = deserializeLogData(logData);
+        } catch (Exception e) {
+            return;
+        }
+
+        if (logEntry.getType() == LogEntry.LogEntryType.SMR) {
+            // Just one stream, always
+            UUID streamId = logData.getStreams().iterator().next();
+            applySmrEntryToStream(streamId, (SMREntry) logEntry);
+        }
+        else if (logEntry.getType() == LogEntry.LogEntryType.MULTIOBJSMR) {
+            MultiObjectSMREntry multiObjectLogEntry = (MultiObjectSMREntry) logEntry;
+            multiObjectLogEntry.getEntryMap().forEach((stream, multiSmrEntry) -> {
+                multiSmrEntry.getSMRUpdates(stream).forEach((smrEntry) -> {
+                    applySmrEntryToStream(stream, smrEntry);
+                });
+            });
+
+        }
+    }
+
+    private void findAndSetLogHead() {
+        logHead = 0;
+    }
+
+    private void findAndSetLogTail() {
+        logTail = runtime.getSequencerView().nextToken(Collections.emptySet(), 0).getTokenValue();
+    }
+
+    /** Initialize log head and log tails
+     *
+     * If logHead and logTail has not been initialized by
+     * the user, initialize to default.
+     *
+     */
+    private void initializeHeadAndTails() {
+        if (logHead < 0) {
+            findAndSetLogHead();
+        }
+
+        if (logTail < 0) {
+            findAndSetLogTail();
+        }
+    }
+
+    /** Book keeping of the the stream tails
+     *
+     * @param logData
+     */
+    public void updateStreamTails(ILogData logData) {
+         for (UUID streamId : logData.getStreams()) {
+             streamTails.put(streamId, logData.getGlobalAddress());
+            }
+    }
+
+    /** Processing logdata at a given log address
+     *
+     * @param logData
+     * @param address
+     */
+    public void processAddressLogData(ILogData logData, long address) {
+        switch (logData.getType()) {
+            case DATA:
+                processLogData(logData);
+                break;
+            case HOLE:
+                break;
+            case TRIMMED:
+                log.warn("loadMaps[{}, start={}]: address is trimmed", address, logHead);
+                // Should not happen
+            case EMPTY:
+                log.warn("loadMaps[address={}]: is empty", address);
+                break;
+            case RANK_ONLY:
+                break;
+            default:
+                break;
+        }
+    }
+
+
+    /** Entry point to load the SMRMaps in memory.
+     *
+     * When this function returns, the maps are fully loaded.
+     *
+     */
+    public void loadMaps() {
+        log.info("loadMaps: Starting to resurrect maps");
+        initializeHeadAndTails();
+
+        long nextRead = 0;
```

Comment:
Can you add a ticker that is increased on every entry that is processed so that you guarantee that the whole address space (head to tail) is processed, I'm worried that there could be off-by-one problems